crates/starknet-replay/src/rpc_state_reader.rs

```diff
@@ -0,0 +1,144 @@
+use blockifier::execution::contract_class::ContractClass;
+use blockifier::state::errors::StateError;
+use blockifier::state::state_api::{StateReader, StateResult};
+use starknet::core::types::BlockId;
+use starknet::providers::jsonrpc::JsonRpcTransport;
+use starknet::providers::{JsonRpcClient, Provider, ProviderError};
+use starknet_api::core::{ClassHash, CompiledClassHash, ContractAddress, Nonce};
+use starknet_api::hash::StarkFelt;
+use starknet_api::state::StorageKey;
+use starknet_os_types::deprecated_compiled_class::GenericDeprecatedCompiledClass;
+use starknet_os_types::hash::GenericClassHash;
+use starknet_os_types::sierra_contract_class::GenericSierraContractClass;
+
+use crate::utils::{execute_coroutine, felt_api2vm, felt_vm2api};
+
+pub struct AsyncRpcStateReader<T>
+where
+    T: JsonRpcTransport,
+{
+    provider: JsonRpcClient<T>,
+    block_id: BlockId,
+}
+
+impl<T> AsyncRpcStateReader<T>
+where
+    T: JsonRpcTransport,
+{
+    pub fn new(provider: JsonRpcClient<T>, block_id: BlockId) -> Self {
+        Self { provider, block_id }
+    }
+}
+
+fn provider_error_to_state_error(provider_error: ProviderError) -> StateError {
+    StateError::StateReadError(provider_error.to_string())
+}
+
+fn to_state_err<E: ToString>(e: E) -> StateError {
+    StateError::StateReadError(e.to_string())
+}
+
+impl<T> AsyncRpcStateReader<T>
+where
+    T: JsonRpcTransport + Sync + Send + 'static,
+{
+    async fn get_storage_at_async(&self, contract_address: ContractAddress, key: StorageKey) -> StateResult<StarkFelt> {
```

Comment:
I would make these `pub` if the `Async` in `AsyncRpcStateReader` is to be publicly meaningful :)

Reply:
Makes sense!